tools/generate-changelog.py

```diff
@@ -0,0 +1,165 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Script is originally copied and modified from
+# https://github.com/snapcore/system-snaps-cicd-tools/blob/0148d51ac8d1668f8ebe142d91673a65e520aa53/workflows/changelog-from-manifest.py
+
+# Script that compares two manifest.yaml files and creates a paragraph
+# that includes the changes for all deb files. These changes are
+# obtained from the debian changelog for the different packages.
+
+import argparse
+import debian.changelog
+import debian.debian_support
+import gzip
+import requests
+import sys
+import yaml
+from collections import namedtuple
+
+# Returns a dictionary from package name to version, using
+# the packages section.
+# manifest_p: path to manifest to load
+def packages_from_manifest(manifest_p):
+    with open(manifest_p) as manifest:
+        manifest_y = yaml.safe_load(manifest)
+
+        pkg_dict = {}
+        for pkg in manifest_y['packages']:
+            # package_name=version
+            pkg_data = pkg.split('=')
+            pkg_dict[pkg_data[0]] = pkg_data[1]
+        return pkg_dict
+
+def package_name(pkg):
+    t = pkg.split(':')
+    return t[0]
+
+def get_changelog_from_file(docs_d, pkg):
+    chl_path = docs_d + '/' + package_name(pkg) + '/changelog.Debian.gz'
+    with gzip.open(chl_path) as chl_fh:
+        return chl_fh.read().decode('utf-8')
+
+def get_changelog_from_url(pkg, new_v):
+    url = 'https://changelogs.ubuntu.com/changelogs/binary/'
+    safe_name = package_name(pkg)
+    if safe_name.startswith('lib'):
+        url += safe_name[0:4]
+    else:
+        url += safe_name[0]
+    url += '/' + safe_name + '/' + new_v + '/changelog'
+    changelog_r = requests.get(url)
+    if changelog_r.status_code != requests.codes.ok:
+        raise Exception('No changelog found in ' + url + ' - status:' +
+                        str(changelog_r.status_code))
+
+    return changelog_r.text
+
+
+# Gets difference in changelog between old and new versions
+# Returns source package and the differences
+def get_changes_for_version(docs_d, pkg, old_v, new_v, indent):
+    # Try to get changelog from file (only option that will work for
+    # ESM packages), otherwise go to changelogs.ubuntu.com.
+    try:
+        changelog = get_changelog_from_file(docs_d, pkg)
+    except Exception:
+        changelog = get_changelog_from_url(pkg, new_v)
+
+    source_pkg = changelog[0:changelog.find(' ')]
+
+    chl = debian.changelog.Changelog(changelog)
+    old_deb_v = debian.debian_support.Version(old_v)
+    for version in chl.get_versions():
+        vc = debian.debian_support.version_compare(old_deb_v, version)
+        if vc >= 0:
+            break
+
+    # Get the changelog chunk since the version older or equal to old_v
+    change_chunk = ''
+    old_change_start = source_pkg + ' (' + version.__str__() + ')'
+    for line in changelog.splitlines():
+        if line.startswith(old_change_start):
```

Comment:
Should not we fail if we do never match it? Do not we risk to have huge changelogs if for some reason the formatting changes.

Reply:
added, good point!